### PR TITLE
[Tech]: Add mechanism for external regulation for assets operations

### PR DESCRIPTION
**Pull request by HaidarJbeily7**

### CODE REVIEW

File: pallets/assets/src/lib.rs [hunk 1]
```diff
@@ -75,7 +75,7 @@ use traits::{
 pub use weights::WeightInfo;
 
 pub type AssetIdOf<T> = <T as Config>::AssetId;
-pub type Permissions<T> = permissions::Pallet<T>;
+// pub type Permissions<T> = permissions::Pallet<T>;
```
Comment (wer1st) [inline pallets/assets/src/lib.rs:78]: delete

File: pallets/assets/src/lib.rs [hunk 2]
```diff
@@ -258,6 +256,8 @@ pub mod pallet {
 
         /// Weight information for extrinsics in this pallet.
         type WeightInfo: WeightInfo;
+
+        type AssetRegulator: AssetRegulator<Self::AccountId, Self::AssetId>;
```
Comment (wer1st) [inline pallets/assets/src/lib.rs:260]: Please place it above of `WeightInfo`

File: common/src/traits.rs [hunk 3]
```diff
@@ -1170,3 +1170,69 @@ impl<AccountId> ReferrerAccountProvider<AccountId> for () {
...
+        issuer: &AccountId,
+        to: Option<&AccountId>,
+        asset_id: &AssetId,
+    ) -> Result<(), DispatchError>;
+
+    /// Check asset regulation for transferring a specific `asset_id` by a specific `account_id`
+    fn transfer(from: &AccountId, to: &AccountId, asset_id: &AssetId) -> Result<(), DispatchError>;
+
+    /// Check asset regulation for minting a specific `asset_id` by a specific `account_id`
+    fn burn(
```
Comment (wer1st) [inline common/src/traits.rs:1192]:
```suggestion
    fn check_mint_permission(
        issuer: &AccountId,
        to: Option<&AccountId>,
        asset_id: &AssetId,
    ) -> Result<(), DispatchError>;

    /// Check asset regulation for transferring a specific `asset_id` by a specific `account_id`
    fn check_transfer_permission(from: &AccountId, to: &AccountId, asset_id: &AssetId) -> Result<(), DispatchError>;

    /// Check asset regulation for minting a specific `asset_id` by a specific `account_id`
    fn check_burn_permission(
```

Comment (wer1st) [reply common/src/traits.rs:1192]: To avoid misunderstanding

File: common/src/traits.rs [hunk 4]
```diff
@@ -1170,3 +1170,69 @@ impl<AccountId> ReferrerAccountProvider<AccountId> for () {
         None
     }
 }
+
+pub trait AssetRegulator<AccountId, AssetId> {
+    /// Assign permission on registering new asset id
+    fn assign_permissions_on_register(
+        owner: &AccountId,
+        asset_id: &AssetId,
+    ) -> Result<(), DispatchError>;
```
Comment (wer1st) [inline common/src/traits.rs:1179]:
```suggestion
    /// Assign permission
    fn assign_permissions(
        owner: &AccountId,
        asset_id: &AssetId,
        permission_id: PermissionId
    ) -> Result<(), DispatchError>;
```

Comment (wer1st) [reply common/src/traits.rs:1179]: It is better to generalize the trait

File: common/src/traits.rs [hunk 5]
```diff
@@ -1170,3 +1170,69 @@ impl<AccountId> ReferrerAccountProvider<AccountId> for () {
...
+    /// Check asset regulation for transferring a specific `asset_id` by a specific `account_id`
+    fn transfer(from: &AccountId, to: &AccountId, asset_id: &AssetId) -> Result<(), DispatchError>;
+
+    /// Check asset regulation for minting a specific `asset_id` by a specific `account_id`
+    fn burn(
+        issuer: &AccountId,
+        from: Option<&AccountId>,
+        asset_id: &AssetId,
+    ) -> Result<(), DispatchError>;
+}
```
Comment (wer1st) [inline common/src/traits.rs:1197]:
```suggestion
    ) -> Result<(), DispatchError>;
    
    /// Check the permission `permission_id` of `issuer` for `asset_id`
    fn check_permission(issuer: &AccountId, asset_id: &AssetId, permission_id: PermissionId) -> Result<(), DispatchError>;
}
```

Comment (wer1st) [reply common/src/traits.rs:1197]: More universal

File: pallets/permissions/src/lib.rs [hunk 6]
```diff
@@ -76,8 +75,8 @@ pub enum Scope {
     Unlimited,
 }
 
-pub const MINT: PermissionId = 2;
-pub const BURN: PermissionId = 3;
+pub const MINT: PermissionId = common::MINT;
+pub const BURN: PermissionId = common::BURN;
 pub const SLASH: PermissionId = 4;
 pub const INIT_DEX: PermissionId = 5;
 pub const MANAGE_DEX: PermissionId = 6;
```
Comment (wer1st) [inline pallets/permissions/src/lib.rs:82]: Don't split permission ids into several files. Select the one place for all permission ids.

File: common/src/constants.rs
Comment (wer1st) [inline common/src/constants.rs:1]: Rename the file to `permissions.rs`

File: common/src/constants.rs [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+// THIS SOFTWARE IS PROVIDED BY Polka Biome Ltd. AS IS AND ANY EXPRESS OR IMPLIED WARRANTIES,
+// INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+// A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL Polka Biome Ltd. BE LIABLE FOR ANY
+// DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+// BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
+// OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+// STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+// USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+use crate::PermissionId;
```
Comment (wer1st) [inline common/src/constants.rs:31]: Move type here
```suggestion
pub type PermissionId = u32;
```

File: common/src/lib.rs [hunk 8]
```diff
@@ -69,6 +70,7 @@ use sp_runtime::TransactionOutcome;
 pub use traits::Config;
 pub mod prelude {
     pub use super::balance_unit::*;
+    pub use super::constants::*;
```
Comment (wer1st) [inline common/src/lib.rs:73]:
```suggestion
    pub use super::permissions;
```

Comment (wer1st) [reply common/src/lib.rs:73]: It's better to use `permissions::MINT` instead of just `MINT`

File: pallets/assets/src/lib.rs [hunk 9]
```diff
@@ -53,11 +53,14 @@ mod mock;
 mod tests;
 
 use codec::{Decode, Encode};
-use common::prelude::{Balance, SwapAmount};
+use common::prelude::{
+    permissions::{BURN, MINT},
```
Comment (wer1st) [inline pallets/assets/src/lib.rs:57]: `common::permissions` without `prelude`

File: pallets/permissions/src/lib.rs [hunk 10]
```diff
@@ -48,24 +48,28 @@
...
 use sp_core::hash::H512;
 use sp_std::vec::Vec;
-
 #[cfg(test)]
 mod mock;
 
 #[cfg(test)]
 mod tests;
 
+pub use common::prelude::permissions::{
```
Comment (wer1st) [inline pallets/permissions/src/lib.rs:65]:
```suggestion
pub use common::permissions::{
```

**Review by wer1st — APPROVED** [review]

**Review by vovac12 — APPROVED** [review]

File: common/src/traits.rs [hunk 11]
```diff
@@ -1444,3 +1446,48 @@ impl<AccountId> ReferrerAccountProvider<AccountId> for () {
         None
     }
 }
+
+pub trait AssetRegulator<AccountId, AssetId> {
```
Comment (Alexey-N-Chernyshov) [inline common/src/traits.rs:1450]: Please, add doc comment for the trait.

**Review by Alexey-N-Chernyshov — APPROVED** [review]

**Pull request merged**